compiler-core/src/recipe.rs

```diff
@@ -0,0 +1,158 @@
+use std::fmt;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use crate::config::PackageConfig;
+use crate::{Error, Result};
+use hexpm::version::Range;
+use serde::de::{self, Deserializer, MapAccess, Visitor};
+use serde::ser::{Serialize, SerializeMap, Serializer};
+use serde::Deserialize;
+
+#[derive(Deserialize, Debug, PartialEq, Eq, Clone)]
+#[serde(untagged, remote = "Self")]
+pub enum Recipe {
+    Hex { version: Range },
+    Path { path: PathBuf },
+    Git { git: String },
+}
+
+impl Recipe {
+    pub fn version_range(&self) -> Result<Range, Error> {
+        match self {
+            Recipe::Hex { version: range } => Ok(range.clone()),
+            Recipe::Path { path } => {
+                let mut config_path = path.clone();
+                config_path.push("gleam.toml");
+                let toml = std::fs::read_to_string(&config_path).map_err(|_| {
+                    Error::DependencyResolutionFailed("Local dependency could not be found".into())
+                })?;
+                let config: PackageConfig = toml::from_str(&toml).map_err(|_| {
+                    Error::DependencyResolutionFailed(
+                        "Local dependency config could not be parsed".into(),
+                    )
+                })?;
+                Ok(Range::new(format!("== {}", config.version.to_string())))
+            }
+            Recipe::Git { .. } => Err(Error::DependencyResolutionFailed(
+                "Git dependencies are currently unsuported".to_string(),
+            )),
+        }
+    }
+
+    pub fn hex(range: &str) -> Recipe {
+        Recipe::Hex {
+            version: Range::new(range.to_string()),
+        }
+    }
+
+    pub fn path(path: &str) -> Recipe {
+        Recipe::Path { path: path.into() }
+    }
+
+    pub fn git(url: &str) -> Recipe {
+        Recipe::Git {
+            git: url.to_string(),
+        }
+    }
+}
+
+// Serialization
+
+impl ToString for Recipe {
+    fn to_string(&self) -> String {
+        match self {
+            Recipe::Hex { version: range } => {
+                format!(r#"{{ version = "{}" }}"#, range.to_string())
+            }
+            Recipe::Path { path } => format!(r#"{{ path = "{}" }}"#, path.display()),
+            Recipe::Git { git: url } => format!(r#"{{ git = "{}" }}"#, url),
+        }
+    }
+}
```

Comment:
What's this used for? I don't think we'd want the ToString trait implemented for these, and this looks specifically like toml serialisation rather than a string representation.

Reply:
This comes from a very frustrating trait impl error somewhere in the manifest that I didn't want to spend time dissecting. I will see if I can move it.

Reply:
I didn't answer your question. Yes this is used for toml serialization, and yes it 100% shouldn't be there.